File: 树/543.二叉树的直径.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 每个节点既要提供当前深度信息(作为一边的最大长度), 还要计算这个点的最大直径
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = [0]
        def dfs(root):
            if not root:
                return -1
            l, r = dfs(root.left), dfs(root.right)
            ans[0] = max(ans[0], l + r + 2)
            return max(l, r) + 1
        dfs(root)
        return ans[0]

File: 树/543.二叉树的直径.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  if (!root) return;
  let res = 0;
  function helper(node) {
    if (!node) return 0;
    const left = helper(node.left);
    const right = helper(node.right);
    res = Math.max(res, left + right + 1);
    return 1 + Math.max(left, right);
  }
  helper(root);
  return res - 1;
};
